### ISSUE TITLE
non-deterministic test failures

### ISSUE THREAD

**Issue by cmelone** [StanfordLegion/legion#1538]
We ran HTR with the `shardrefine` branch at the request of @lightsighter. Our regression tests are randomly failing their tolerance checks in debug mode. I don't have much useful info to share at the moment, but @mariodirenzo will be doing some analysis with legion spy to get some clues.

@elliottslaughter, could you please add this to #1032 ?

**Comment by cmelone** [comment]
I've done some additional testing and found that this issue is not specific to `shardrefine`. It was difficult to identify in our CI because it only seems to fail one test about 1/20th of the time.

This regression was introduced between the `48d22ec` and `149be33` commits of `control_replication`. Are there any commits within those two that would be the most suspicious? We can then revert them and see if the regression disappears, as opposed to iterating through each one.

Thanks

**Comment by lightsighter** [comment]
You're saying that `48d22ec` works correctly all the time or that is the first commit that you notice that fails? 

Unfortunately, most of the changes in that commit range come from Realm so we're probably going to need to binary search, but none of the Realm changes jump out as immediately scary to me. There's actually only three Legion changes in that range: one adding support for initial futures in future map all-reduces (which you don't use), one adding checks that users don't mix color allocation modes for partitions (which would fail deterministically if you hit it), and one improving the Legion shutdown protocol (which won't impact any running code). If you want to check any of those Legion changes, the one to remove is `f8d66ead` as it's possible that influences a future reduction somewhere.

**Comment by cmelone** [comment]
48d22ec is the first one that fails. I’ll try your suggestion, thanks

**Comment by lightsighter** [comment]
Ok, if `48d22ec` is the first one that fails we need to go back further to properly bound the set of commits that can be the cause.

**Comment by lightsighter** [comment]
`48d22ec` is a merge commit that merges in several changes from master. Keep going backwards through the commits until you find a commit that is stable, e.g. the next one back is probably `5f8bc7b4`.

**Comment by mariodirenzo** [comment]
Mike, `5f8bc7b4` had issues that we fixed together. The code does not even run on that commit.
As written by @cmelone, `149be33`, which is the commit just before `5f8bc7b4`, works fine. It must be something between `149be33` and `48d22ec` and we can't bisect further. We need to start reverting specific sets of commits.

**Comment by lightsighter** [comment]
> As written by @cmelone, 149be33, which is the commit just before 5f8bc7b4, works fine

I don't think that's accurate. The order of commits from oldest to newest is:

5f8bc7b - August 15th
48d22ec - August 15th
149be33 - August 18th

`48d22ec` comes between `5f8bc7b` and `149be33`.  And @cmelone said:

> 48d22ec is the first one that fails.

But if it is also failing that means it's possible that is not the first one that fails. It's possible that the failure was introduced even earlier. 

> Mike, 5f8bc7b4 had issues that we fixed together. The code does not even run on that commit.

You may need to cherry-pick in the Realm changes to the machine model parsing to determine if the cause of the non-determinism occurred even earlier than `48d22ec`.

**Comment by mariodirenzo** [comment]
Sorry, I misread your comment. I thought you were proposing to bisect at [f8d66ea](https://gitlab.com/StanfordLegion/legion/-/commit/f8d66ead0d032cee04dfa80d7aeb7f8b92b9bec1).
Last time I chatted with @cmelone, I understood that   [48d22ec](https://github.com/StanfordLegion/legion/commit/48d22ecfec1c66c6c9f31dd18600eb4e086c89b5) works fine and that [149be33](https://gitlab.com/StanfordLegion/legion/-/commit/149be337863232061201dfea27a8a7a794d338dc) fails

**Comment by cmelone** [comment]
Reverting `f8d66ead1` solves the issue...

**Comment by lightsighter** [comment]
> Reverting f8d66ead1 solves the issue...

That commit comes after `48d22ec`, but you said that `48d22ec` was failing also...

**Comment by cmelone** [comment]
@lightsighter apologies, I checked again and `48d22ec` succeeds.

**Comment by lightsighter** [comment]
@cmelone This commit adds support for an initial value for future reductions in index task launches. Are you actually using that functionality? I'd be pretty surprised if this was actually the commit that is the first one that fails as it should all result in a no-op if you're not even using that feature. How many times do you do the runs when checking a commit? Is it always on the same allocation of nodes?

**Comment by cmelone** [comment]
Yes, same allocation of nodes and I was running the same execution 50 times per commit. However, I tried a more recent commit (`1efeabc`) and I can't reproduce after running >150 times. Something must have changed since `fd5f489` (the last commit that failed). If you'd like, I can close this, or I can do some more testing over the weekend to know the root cause of this issue, if we run into it in the future.

**Comment by lightsighter** [comment]
I think you should do some more testing to convince yourself that the issue is actually fixed. Certainly bugs were found and fixed in the intermediate time frame, but none that sound like the failure mode that we have here where things were silently getting the wrong answer. It's up to you guys to decide when you want to close the issue, but it's hard to say if we fixed the issue if we never understood the root cause.

**Comment by mariodirenzo** [comment]
I agree with Mike, let's keep this issue around.
I do not think that we should keep doing active testing specifically for this issue but we keep an eye on our CI and see if it starts failing again with the same mode.

**Comment by cmelone** [comment]
I have not seen this reappear in our CI, but I went back and did some additional testing to find out what the root cause might be. It would probably be easier to disregard the earlier commits discussed on this thread for simplicity's sake.

Commit [`39a6b0ee26a0a077e0fc37d5e1654ebedaf89815` ](https://gitlab.com/StanfordLegion/legion/-/commit/39a6b0ee26a0a077e0fc37d5e1654ebedaf89815)fails repeatedly, but I have ran our test case 1200 times on commit [`039d79a60f65307d0533b595bd21f5e5ccc6c44c`](https://gitlab.com/StanfordLegion/legion/-/commit/039d79a60f65307d0533b595bd21f5e5ccc6c44c) and it has only succeeded. 

Does this make sense in context of the failure mode? (the simulation giving erroneous output and failing our tolerance checks).

**Comment by lightsighter** [comment]
I did fix at least one bug with execution fences in that second commit. I don't see why the first commit would have caused any problems though as it was exclusively about the padded instances (usually their failure mode would be an illegal memory access). It sounds though like you've done some more rigorous testing though now, which at least should improve our confidence that we fixed whatever the error is.
